### ISSUE TITLE
Unexpected outcome with OR clause

### ISSUE THREAD

**Issue by shubh9194** [vespa-engine/vespa#23599]
**Describe the bug**
We have a cluster with 1 config server and 18 content nodes in a group distribution with almost 40 million document
we are trying to run some queries on it but OR clause giving some unexpected outcome.

**To Reproduce**
First Ran these query
```
 "select * from sources * where (table contains \"com.test.document.store.lex\" AND weakAnd(party_name_s contains \"Al Holbrook\",assignedorg_s contains ([{\"weight\": 1000}]\"FF\"), disclosure_type_s contains \"Quantitative Investment Strategies\", accountname_s contains \"DOUGLAS G. SMITH\", document_access_method_s contains ([{\"weight\":1000}]\"CUSTOM\"), transfereeaccount_s contains \"EA0326627\") AND weakAnd(documentactive_s contains ([{\"weight\":1000}]\"false\"),documentactive_s contains \"true\") AND weakAnd(assetclass_s contains \"PRIVATE_EQUITY\",assetclass_s contains\"EXCHANGE_FUND\") AND weakAnd(documentcode_s contains ([{\"weight\": 1000}]\"APPXA\"),documentcode_s contains \"SUBDOC\") AND weakAnd(signed_s contains ([{\"weight\": 1000}]\"true\"),signed_s contains \"false\") AND weakAnd(ordertype_s contains \"BUY\", saleschannel_s contains \"PWM\"))";
```

output :
{
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 6877
        }

then added more clauses on above query.
```
"select * from sources * where (table contains \"com.test.document.store.lex\" AND weakAnd(party_name_s contains \"Al Holbrook\",assignedorg_s contains ([{\"weight\": 1000}]\"FF\"), disclosure_type_s contains \"Quantitative Investment Strategies\", accountname_s contains \"DOUGLAS G. SMITH\", document_access_method_s contains ([{\"weight\":1000}]\"CUSTOM\"), transfereeaccount_s contains \"EA0326627\") AND weakAnd(documentactive_s contains ([{\"weight\":1000}]\"false\"),documentactive_s contains \"true\") AND weakAnd(assetclass_s contains \"PRIVATE_EQUITY\",assetclass_s contains\"EXCHANGE_FUND\") AND weakAnd(documentcode_s contains ([{\"weight\": 1000}]\"APPXA\"),documentcode_s contains \"SUBDOC\") AND weakAnd(signed_s contains ([{\"weight\": 1000}]\"true\"),signed_s contains \"false\") AND weakAnd(ordertype_s contains \"BUY\", saleschannel_s contains \"PWM\"))  AND (collection contains \"com.test.document.store.lex.content\") ;"
```
output:
{
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 6400
        },

Then added an OR clause on collection fields
```
 "select * from sources * where (table contains \"com.test.document.store.lex\" AND weakAnd(party_name_s contains \"Al Holbrook\",assignedorg_s contains ([{\"weight\": 1000}]\"FF\"), disclosure_type_s contains \"Quantitative Investment Strategies\", accountname_s contains \"DOUGLAS G. SMITH\", document_access_method_s contains ([{\"weight\":1000}]\"CUSTOM\"), transfereeaccount_s contains \"EA0326627\") AND weakAnd(documentactive_s contains ([{\"weight\":1000}]\"false\"),documentactive_s contains \"true\") AND weakAnd(assetclass_s contains \"PRIVATE_EQUITY\",assetclass_s contains\"EXCHANGE_FUND\") AND weakAnd(documentcode_s contains ([{\"weight\": 1000}]\"APPXA\"),documentcode_s contains \"SUBDOC\") AND weakAnd(signed_s contains ([{\"weight\": 1000}]\"true\"),signed_s contains \"false\") AND weakAnd(ordertype_s contains \"BUY\", saleschannel_s contains \"PWM\"))  AND (  collection contains \"com.test.document.store.lex.content\" OR collection contains \"com.gs.imd.activities.document.fi_investmentview\") ;"
```
output
{
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 0
 }

Don't understand why it returns 0 documents after adding the OR clause.

**Expected behavior**
It should have returned the same number documents after adding the OR clause even if the new condition does not match with any document.

**Vespa version**
7.589.25

**Additional context**

Update : removing some company specific terms from queries

**Comment by bjormel** [comment]
To see how the YQL expression you run is handled, you can add the request parameter [trace.level](https://docs.vespa.ai/en/reference/query-api-reference.html#trace.level) to the request.

You can see an example of the output from the trace using `trace.level=2` here: https://docs.vespa.ai/en/query-api.html#query-tracing 

If you are not able to get further with this, please include the trace so we can help you debug.

Thanks!

**Comment by bratseth** [comment]
Your queries look fine. I guess the last query probably timed out from matching a log of documents.
You can see this by looking at the coverage information in the response, and you can increase timeout by passing e.g &timeout=5s

**Comment by shubh9194** [comment]
@bratseth, I will try this and update you further

**Comment by shubh9194** [comment]
[traceLevelAfterOR.txt](https://github.com/vespa-engine/vespa/files/9359447/traceLevelAfterOR.txt)
[traceLevelBeforeOR.txt](https://github.com/vespa-engine/vespa/files/9359448/traceLevelBeforeOR.txt)

Attached the tracelevel Output for 2nd and 3rd query in description

**Comment by bratseth** [comment]
Thank you. 

tl;dr: Use OR instead of WeakAnd for your structured conditions.

The query here is

```
AND 
table:'com gs pwm document stor lex' 
(WEAKAND(100) party_name_s:'al holbrok' assignedorg_s:ff!1000 disclosure_type_s:'quantitativ investment strategies' accountname_s:'douglas g smith' document_access_method_s:custom!1000 transfereeaccount_s:ea0326627) 
(WEAKAND(100) documentactive_s:fal!1000 documentactive_s:true) 
(WEAKAND(100) assetclass_s:'privat equity' assetclass_s:'exchang fund') 
(WEAKAND(100) documentcode_s:appxa!1000 documentcode_s:subdoc) 
(WEAKAND(100) signed_s:true!1000 signed_s:fal) 
(WEAKAND(100) ordertype_s:buy saleschannel_s:pwm) 
(OR collection:'com gs pwm document stor lex content' collection:'com gs imd activities document fi investmentview')
```
... in the _after OR_ case, in the _before OR_ case the last line is just the first of the two items.

Each WEAKAND(100) operator will return at least 100 hits matching some of the items of it (per node).
Because of this, AND'ing two WEAKAND operators may produce 0 matches, if those two operators happen to produce disjunct sets of hits.

If the other filters in the query are weaker, the probability of that goes up. Adding the OR makes that filter weaker.

Looking at the content of these items, I think these are structured data which should act as filters and you want to OR them to surface all the matches. I guess that's where you started and this is an attempt at optimization? It becomes faster due to discarding matches locally, but unfortunately that also means you can't expect global correctness, although you can of course also increase the probability by using a larger N than 100.

If match details are not accessed in ranking, it would increase performance a lot to mark them as _filter_ terms. If you haven't I would try that first.

**Comment by shubh9194** [comment]
@bratseth, in the after OR case, in the before OR case the last line is just the first of the two items.
I don't understand what you mean by above statement

**Comment by shubh9194** [comment]
Here tale the example of these simple queries
"select * from sources * where (collection contains \"com.gs.pwm.document.store.lex.content\") ;"
this returns "totalCount": 1038307

"select * from sources * where (collection contains \"com.gs.imd.activities.document.fi_investmentview\" OR collection contains \"com.gs.pwm.document.store.lex.content\") ;"
this returns "totalCount": 3107
But this query should have return all documents that matches either of the condition In OR and should have return large number of documents.

I am unable to understand the behavior here
now the second query: "select * from sources * where (collection contains \"com.gs.imd.activities.document.fi_investmentview\") ;"
this returns "totalCount": 3107

Now OR the condition in above two queries into single query

**Comment by bratseth** [comment]
By "in the after OR case, in the before OR case the last line is just the first of the two items." I was just summarizing the difference between the two queries you shared.

In your last comment you provide two queries which lack all the WEAKAND clauses of the queries you asked about earlier. Are you saying you observe the same thing running these much simpler queries. If so, please share the traces of _those_ queries instead of the more complex ones discussed above.

**Comment by shubh9194** [comment]
[tracelevel_2ndQuery.txt](https://github.com/vespa-engine/vespa/files/9464692/tracelevel_2ndQuery.txt)
[tracelevelWithOR_3rdQuey.txt](https://github.com/vespa-engine/vespa/files/9464693/tracelevelWithOR_3rdQuey.txt)

Attached the tracelevel output

**Comment by bratseth** [comment]
Okay, both those queries matches the same number of documents (3107), none contains any WeakAnds, and none time out.
Since the second is equal to the first plus an additional OR term, that indicates the second term does not have any other matches, and at most 3107 matches if evaluated standalone. Can you also share the query that shows this is not the case? Also: Check if you have match-phase settings in the rank profile. That will also allow adaptively rejecting matches.

**Comment by shubh9194** [comment]
"select * from sources * where (collection contains "com.gs.pwm.document.store.lex.content") ;"
this returns "totalCount": 1038307

This is the result when we use only 2nd term of OR condition so we do have matches for this and it is lot more that 1st term.  I will also attach the tracelevel for this.

**Comment by shubh9194** [comment]
@bratseth Any update on this?

**Comment by geirst** [comment]
@shubh9194 : Can you please also share more detailed backend trace information for the queries used in tracelevel_2ndQuery.txt, tracelevelWithOR_3rdQuey.txt and tracelevel_query1.txt, by also adding "&trace.explainLevel=1" to the queries? That will make us understand how these queries are executed on the content nodes.

**Comment by shubh9194** [comment]
[tracelevel_query1.txt](https://github.com/vespa-engine/vespa/files/9624302/tracelevel_query1.txt)
[tracelevel_query2.txt](https://github.com/vespa-engine/vespa/files/9624303/tracelevel_query2.txt)
[tracelevelWithOR_3rdQuery.txt](https://github.com/vespa-engine/vespa/files/9624304/tracelevelWithOR_3rdQuery.txt)

@geirst please find the attached tracelevel generated with following parameters
    "timeout": "120s",
    "hits": 0,
    "tracelevel":2,
    "trace.explainLevel":1

**Comment by geirst** [comment]
It turns out the information we need are not part of trace information using those levels on the old Vespa 7 version. @shubh9194 can you please run the three queries again using "&tracelevel=6&trace.explainLevel=1" ?

**Comment by geirst** [comment]
@shubh9194 : After analysing the trace information, we have found that the problem is caused by different stemming in the queries.

In query 1 the language is detected to be ENGLISH and 'store' is stemmed to 'store' (unchanged).
```
{
    "message": "Stemming with language=ENGLISH"
},
{
    "message": "Stemming: [\nselect * from sources * where collection contains ([{origin: {original: \"com.gs.pwm.document.store.lex.content\", offset: 0, length: 37}}]phrase(\"com\", \"gs\", \"pwm\", \"document\", \"store\", \"lex\", \"content\")) ...
},
...
{
    "message": "sc0.num0 search to dispatch: query=[collection:'com gs pwm document store lex content'] timeout=119998ms offset=0 hits=0 groupingSessionCache=true sessionId=2f62bb90-f462-46eb-8047-9a61a8fc0eb1.1663923489889.343.default grouping=0 :  restrict=[imdamdeloreanlexDocument]"
},
```

But in query 3 the language is detected to be DUTCH and 'store' is stemmed to 'stor'. The following phrase (collection:'com gs pwm document stor lex content') is sent to the backend and does not match any documents. Only the 3107 hits from the other phrase under the OR are returned.
```
{
    "message": "Stemming with language=DUTCH"
},
{
    "message": "Stemming: [\nselect * from sources * where (collection contains ([{origin: {original: \"com.gs.imd.activities.document.fi_investmentview\", offset: 0, length: 48}}]phrase(\"com\", \"gs\", \"imd\", \"activities\", \"document\", \"fi\", \"investmentview\")) OR collection contains ([{origin: {original: \"com.gs.pwm.document.store.lex.content\", offset: 0, length: 37}}]phrase(\"com\", \"gs\", \"pwm\", \"document\", \"stor\", \"lex\", \"content\"))) ...
},
...
{
    "message": "sc0.num0 search to dispatch: query=[OR collection:'com gs imd activities document fi investmentview' collection:'com gs pwm document stor lex content'] timeout=119990ms offset=0 hits=0 groupingSessionCache=true sessionId=2f62bb90-f462-46eb-8047-9a61a8fc0eb1.1663923195482.190.default grouping=0 :  restrict=[imdamdealflowlexDocument]"
},
```

To fix this you need to specify in the query request which language to use, e.g. "&model.language=en". For more details see https://docs.vespa.ai/en/linguistics.html#querying-with-language and https://docs.vespa.ai/en/reference/query-api-reference.html#model.language. FYI @bratseth.

**Comment by shubh9194** [comment]
Ok I got it but I have another doubt.
when I add model.language=en to second query, it returns 0 document now instead of 3107
[traceLevel_query2_en.txt](https://github.com/vespa-engine/vespa/files/9634860/traceLevel_query2_en.txt)

**Comment by geirst** [comment]
In the original query 2 the language was also detected to be DUTCH. The phrase sent to the backend was unchanged: (collection:'com gs imd activities document fi investmentview'). Apparently, this matches with how these documents where stemmed during feeding.
```
{
    "message": "Stemming with language=DUTCH"
},
{
    "message": "Stemming: [\nselect * from sources * where collection contains ([{origin: {original: \"com.gs.imd.activities.document.fi_investmentview\", offset: 0, length: 48}}]phrase(\"com\", \"gs\", \"imd\", \"activities\", \"document\", \"fi\", \"investmentview\")) ...
},
...
{
    "message": "sc0.num0 search to dispatch: query=[collection:'com gs imd activities document fi investmentview'] ...  restrict=[amCRPlexDocument]"
},
```

When using "model.language=en" however, we observe that the stemming is different, and the phrase sent to the backend is instead (collection:'com gs imd **activity** document fi investmentview'), which doesn't match.
```
{
    "message": "Stemming with language=ENGLISH"
},
{
    "message": "Stemming: [\nselect * from sources * where collection contains ([{origin: {original: \"com.gs.imd.activities.document.fi_investmentview\", offset: 0, length: 48}}]phrase(\"com\", \"gs\", \"imd\", \"activity\", \"document\", \"fi\", \"investmentview\")) ...
},
...
{
    "message": "sc0.num0 search to dispatch: query=[collection:'com gs imd activity document fi investmentview'] ...  restrict=[imdamdealflowlexDocument]"
},
```

It is important that the stemming used during feeding / indexing is the same as the one used at query time. How is the field _collection_ defined in the schema?

**Comment by shubh9194** [comment]
This is how we define the collection fields
field collection type array<string> {
            indexing: summary | index

        }

**Comment by geirst** [comment]
Based on the content of the _collection_ field (some kind of id), we suggest using _match: word_  (https://docs.vespa.ai/en/reference/schema-reference.html#match). This should fix the stemming asymmetry currently observed between feeding / indexing and querying. Note that such change requires re-feeding of all data.
```
field collection type array {
    indexing: summary | index
    match: word
}
```

**Comment by jobergum** [comment]
Which also will significantly improve search performance, avoiding phrase matching. 

This is covered [here](https://docs.vespa.ai/en/performance/feature-tuning.html#indexing-uuids) . Using `match:word` disables stemming and avoids doing phrase search. 

If you know the language of your documents, you should consider overriding language detection, see [linguistics in vespa](https://docs.vespa.ai/en/linguistics.html#indexing-with-language)

**Comment by shubh9194** [comment]
@geirst 
but with above solution, it sees we are unable to run queries having wildcard like
"select * from sources * where (collection contains "com.*") ;"

Is there any other alternate way?

**Comment by jobergum** [comment]
Wildcard or prefix search has never been supported for string fields with `index`, the above will simply search for 'com'

**Comment by geirst** [comment]
If you need prefix matching you must change the field from `index` to `attribute`, and use prefix syntax in the query: https://docs.vespa.ai/en/reference/query-language-reference.html#prefix.
```
field collection type array<string> {
    indexing: summary | attribute
    match: word
}
```

**Comment by shubh9194** [comment]
@jobergum 
I have tried with attribute field as well
select * from sources * where (collection_attr matches \"com.gs.pwm.document.store.lex.content\") ;
this query return 21992 docs

select * from sources * where (collection_attr matches \"com.*.document.store.lex.content\") 
this query also works and returns 21992 docs

but following query return 0 documents
select * from sources * where (collection_attr matches \"com*.document.store.lex.content\") ;
select * from sources * where (collection_attr matches \"com.gs.pwm.docu*.store.lex.content\") ;
select * from sources * where (collection_attr matches \"com*.=document.store*content\") ;
not sure why wildcard does not work for these queries


Also how many terms can we use with OR? Is there any limit?

**Comment by jobergum** [comment]
First of all when changing from `index` to `attribute` you must follow:

- https://docs.vespa.ai/en/schemas.html#schema-modifications
- https://docs.vespa.ai/en/reference/schema-reference.html#modifying-schemas

Then you use `matches` which is r[egular expressions](https://docs.vespa.ai/en/reference/query-language-reference.html#matches), to see how Vespa parses the regular expressions, use `tracelevel` to trace the interpretation.
